Add Advanced Security Alert tools

nuthanmunaiah opened the pull request:
We are introducing two new tools---`advsec_get_alerts` and 
`advsec_get_alert_details`---in this change to allow customers to
work with [GHAzDO (GitHub Advanced Security for Azure
DevOps)](https://aka.ms/advanced-security) alerts directly in code 
editor.

## GitHub issue number

N/A

## **Associated Risks**

None

## ✅ **PR Checklist**

- [x] **I have read the [contribution guidelines](https://github.com/microsoft/azure-devops-mcp/blob/main/CONTRIBUTING.md)**
- [x] **I have read the [code of conduct guidelines](https://github.com/microsoft/azure-devops-mcp/blob/main/CODE_OF_CONDUCT.md)**
- [x] Title of the pull request is clear and informative.
- [x] 👌 Code hygiene
- [ ] 🔭 ~~Telemetry added, updated, or N/A~~ N/A
- [x] 📄 Documentation added, updated, or N/A
- [x] 🛡️ Automated tests added, or N/A

## 🧪 **How did you test it?**

  * Automated unit tests run using `npm run test`
  * Manual testing with VS Code Agent Mode in Copilot Chat.

danhellem commented:
@nuthanmunaiah we don't yet have an owner for the GHAzDO area and may not want to merge anything around that space until we do. Stand by.

danhellem commented:
@nuthanmunaiah also, in the future. Please follow the contribution guidelines before submitting a PR.

https://github.com/microsoft/azure-devops-mcp/blob/main/CONTRIBUTING.md#-writing-code

laurajjiang commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,236 @@
+        .optional()
+        .describe("Filter alerts by validity status. Only applicable for secret alerts."),
+      top: z.number().optional().default(100).describe("Maximum number of alerts to return. Defaults to 100."),
+      orderBy: z.enum(["id", "firstSeen", "lastSeen", "fixedOn", "severity"]).optional().default("id").describe("Order results by specified field. Defaults to 'id'."),
```
Can we default order by severity instead? I don't think most customers would sort by id, nor do we by default.

laurajjiang commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,236 @@
+      toolName: z.string().optional().describe("Filter alerts by tool name."),
+      ref: z.string().optional().describe("Filter alerts by git reference (branch). If not provided and onlyDefaultBranch is true, only includes alerts from default branch."),
+      onlyDefaultBranch: z.boolean().optional().default(true).describe("If true, only return alerts found on the default branch. Defaults to true."),
+      confidenceLevels: z
```
Can we also add that by default, we fetch both? This has been a point of annoyance from some customers that we only return high confidence alerts.

nuthanmunaiah replied:
I noticed that the bug applies to every filter criterion that allows multiple values. For instance, if I want to reason on fixed alerts, the query parameter for state is getting set to `criteria.states.0=4` (`4` here is fixed) instead of `criteria.states=4`. Let me reach out to `azure-devops-node-api` Team to figure out if there is anything we can do to mitigate this.

nuthanmunaiah replied:
Opened microsoft/azure-devops-node-api#644 to fix the bug.

nuthanmunaiah commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,236 @@
+        .array(z.enum(["active", "inactive", "none", "unknown"]))
+        .optional()
+        .describe("Filter alerts by validity status. Only applicable for secret alerts."),
+      top: z.number().optional().default(100).describe("Maximum number of alerts to return. Defaults to 100."),
```
How does one know if the Copilot context window is full? I used the server to get all 549 active dependency scanning alerts for one of our test repositories, and Copilot was able to summarize the results just fine.

laurajjiang replied:
If the tool call is successfully processed and summarized after the call, then we should be good. In my testing, the tool call would loop because it surpassed the context window. Sounds like this was patched/changed in some way.

nuthanmunaiah commented:
@polatengin I added the two functions you recommended to `src/utils.ts`. I also added unit tests to `test/src/utils.test.ts` to test these two new functions..

Note: While I was working on the change, it appears two new similar functions (`getEnumKeys` and `safeEnumConvert`) were added to `src/utils.ts`. `safeEnumConvert` function, however, does not transform an enumeration to a `Record<string, T[keyof T]>` before attempting to map a string to an enumeration value. A side effect of this implementation choice is that the `key` argument to `safeEnumConvert` must be case sensitive.

**Example**

```TypeScript
enum AlertType { Unknown = 0, Dependency = 1, Secret = 2, Code = 3, License = 4 }

console.log(safeEnumConvert(AlertType, "Code")); // Returns `3`
console.log(safeEnumConvert(AlertType, "code")); // Returns `undefined`
```

Novaes commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,139 @@
+      repository: z.string().describe("The name or ID of the repository to get alerts for."),
+      alertType: z.enum(["code", "dependency", "secret", "unknown"]).optional().describe("Filter alerts by type. If not specified, returns all alert types."),
+      states: z
+        .array(z.enum(["active", "dismissed", "fixed", "autoDismissed", "unknown"]))
```
Can we use the typescript sdk enums here?  We have some examples on repos.

Novaes commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,139 @@
+    {
+      project: z.string().describe("The name or ID of the Azure DevOps project."),
+      repository: z.string().describe("The name or ID of the repository to get alerts for."),
+      alertType: z.enum(["code", "dependency", "secret", "unknown"]).optional().describe("Filter alerts by type. If not specified, returns all alert types."),
```
same here for using typescript sdk enums

Novaes commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,139 @@
+        .optional()
+        .describe("Filter alerts by state. If not specified, returns alerts in any state."),
+      severities: z
+        .array(z.enum(["critical", "high", "medium", "low", "note", "error", "warning", "undefined"]))
```
same for those

Novaes commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,139 @@
+        const connection = await connectionProvider();
+        const alertApi = await connection.getAlertApi();
+
+        const criteria = {
```
This is: 
* Mapping string inputs to enum values using mapStringToEnum(...).
* Filtering out undefined values to ensure only valid enum entries are included.
* Conditionally including fields only if their corresponding input exists.

this is quite complex, can we move this to a separate class /utils or even at least just a function, to make it easy to maintain this piece of logic.

Novaes commented on src/tools/advsec.ts:
```diff
@@ -0,0 +1,139 @@
+        .array(z.enum(["active", "inactive", "none", "unknown"]))
+        .optional()
+        .describe("Filter alerts by validity status. Only applicable for secret alerts."),
+      top: z.number().optional().default(100).describe("Maximum number of alerts to return. Defaults to 100."),
```
Can you confirm pagination working properly? If yes, I'd even start with lower number, 50 for example.

Novaes commented on src/utils.ts:
```diff
@@ -7,6 +7,21 @@ export const apiVersion = "7.2-preview.1";
 export const batchApiVersion = "5.0";
 export const markdownCommentsApiVersion = "7.2-preview.4";
 
+export function createEnumMapping<T extends Record<string, string | number>>(enumObject: T, keyTransform?: (key: string) => string): Record<string, T[keyof T]> {
```
this seems quite error prone and complex. Can we simplify it? An example: https://stackoverflow.com/questions/41308123/map-typescript-enum

nuthanmunaiah replied:
I simplified it by removing the `keyTransform` argument. The implementation of the `createEnumMapping` is largely inspired by [earlier comment](https://github.com/microsoft/azure-devops-mcp/pull/270#pullrequestreview-3040271177) from @polatengin.

Novaes commented on test/src/tools/advsec.test.ts:
```diff
@@ -0,0 +1,237 @@
+    });
+  });
+
+  describe("advsec_get_alerts tool", () => {
```
missing testing with multiple alerts

nuthanmunaiah replied:
Added test for multiple alerts.

Novaes reviewed (approved):
LGTM with comments 👍

nuthanmunaiah commented on test/src/utils.test.ts:
```diff
@@ -0,0 +1,189 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
```
I am working on a demo of the end-to-end experience of using the two new tools. I will share the recording here after I have had someone review the video to strip any Microsoft-internal content from the video.

nuthanmunaiah commented on test/src/utils.test.ts:
```diff
@@ -0,0 +1,189 @@
+    });
+
+    describe("edge cases", () => {
+      it("should handle undefined input", () => {
```
We test the following two cases when `value` to map to an enumeration is `undefined`:

  1. When there is no `defaultValue` parameter specified, `undefined` should be returned.
  2. When a `defaultValue` of `AlertType.Unknown` is specified, `AlertType.Unknown` should be returned.

Does this help?
